### PR TITLE
Improve perf of Utf8Parser.TryParse for int64 and uint64

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Somewhat inspired by investigating https://github.com/dotnet/runtime/issues/52314 and seeing a little bit of low-hanging fruit that we can address.

Perf results (basically from running [this benchmark](https://github.com/dotnet/performance/blob/716f16c4ce263d03e6fcf1ea8e60bcf5fcbee600/src/benchmarks/micro/libraries/System.Buffers/Utf8ParserTests.cs#L22) with some additional inputs):

|        Method | Toolchain |                value |     Mean |     Error |   StdDev | Ratio |
|-------------- |---------- |--------------------- |---------:|----------:|---------:|------:|
| TryParseInt64 |  baseline |                   -1 |  3.73 ns |  0.022 ns | 0.018 ns |  1.00 |
| TryParseInt64 |   compare |                   -1 |  3.25 ns |  0.025 ns | 0.021 ns |  0.87 |
|               |           |                      |          |           |          |       |
| TryParseInt64 |  baseline |               -12345 |  5.93 ns |  1.680 ns | 0.092 ns |  1.00 |
| TryParseInt64 |   compare |               -12345 |  5.85 ns |  1.097 ns | 0.060 ns |  0.99 |
|               |           |                      |          |           |          |       |
| TryParseInt64 |  baseline |   -12345678901234567 | 15.75 ns | 43.214 ns | 2.368 ns |  1.00 |
| TryParseInt64 |   compare |   -12345678901234567 | 13.24 ns |  0.371 ns | 0.020 ns |  0.85 |
|               |           |                      |          |           |          |       |
| TryParseInt64 |  baseline | -9223372036854775808 | 19.35 ns |  0.326 ns | 0.305 ns |  1.00 |
| TryParseInt64 |   compare | -9223372036854775808 | 17.12 ns |  0.085 ns | 0.076 ns |  0.88 |
|               |           |                      |          |           |          |       |
| TryParseInt64 |  baseline |                    0 |  3.17 ns |  0.097 ns | 0.005 ns |  1.00 |
| TryParseInt64 |   compare |                    0 |  3.21 ns |  0.602 ns | 0.033 ns |  1.01 |
|               |           |                      |          |           |          |       |
| TryParseInt64 |  baseline |                    1 |  3.27 ns |  0.060 ns | 0.003 ns |  1.00 |
| TryParseInt64 |   compare |                    1 |  3.21 ns |  0.322 ns | 0.017 ns |  0.98 |
|               |           |                      |          |           |          |       |
| TryParseInt64 |  baseline |                12345 |  5.76 ns |  0.738 ns | 0.040 ns |  1.00 |
| TryParseInt64 |   compare |                12345 |  5.07 ns |  0.929 ns | 0.050 ns |  0.88 |
|               |           |                      |          |           |          |       |
| TryParseInt64 |  baseline |    12345678901234567 | 13.90 ns |  0.247 ns | 0.013 ns |  1.00 |
| TryParseInt64 |   compare |    12345678901234567 | 12.85 ns |  1.504 ns | 0.082 ns |  0.92 |
|               |           |                      |          |           |          |       |
| TryParseInt64 |  baseline |  9223372036854775807 | 17.72 ns |  0.102 ns | 0.095 ns |  1.00 |
| TryParseInt64 |   compare |  9223372036854775807 | 16.11 ns |  0.063 ns | 0.059 ns |  0.91 |

Some of the benchmarks were going a bit screwy on my machine, so I basically ran them one at a time and concatenated all of the results into a single table. Not sure what was up with my environment.

The tl;dr is that this shouldn't have any worse perf than the existing code, and the total codegen ends up being __448 bytes smaller__ (on x64) compared to baseline across the two methods.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs [hunk 1]
```diff
@@ -336,107 +336,121 @@ private static bool TryParseInt32D(ReadOnlySpan<byte> source, out int value, out
...
-            else if (source[0] == '+')
-            {
-                indexOfFirstDigit = 1;
 
-                if (source.Length <= indexOfFirstDigit)
+                idx++;
+
+                if ((uint)firstChar == unchecked((uint)('-' - '0')))
+                {
+                    sign--; // set to -1
```
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs:367]:
```suggestion
                    sign = -1;
```
Make it explicit?
The JIT will emit code like `mov rax, 0xffffffffffffffff` anyway.

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs [hunk 2]
```diff
@@ -336,107 +336,121 @@ private static bool TryParseInt32D(ReadOnlySpan<byte> source, out int value, out
...
-            value = ((long)parsedValue) * sign;
+            // 'sign' is 0 for non-negative and -1 for negative. This allows us to perform
+            // cheap arithmetic + bitwise operations to mimic a multiplication by 1 or -1
+            // without incurring the cost of an actual multiplication operation.
+            //
+            // If sign = 0,  this becomes value = (parsedValue ^  0) -   0  = parsedValue
+            // If sign = -1, this becomes value = (parsedValue ^ -1) - (-1) = ~parsedValue + 1 = -parsedValue
+
+            bytesConsumed = idx;
+            value = ((long)parsedValue ^ sign) - sign;
```
Comment (gfoidl) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs:447]: 🚀 

The same could be done to other types too (like int32 above, etc.)?

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs (hunk 1 above)
Comment (gfoidl) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Signed.D.cs:367]: So I'd keep it as is. Thanks for the info.

**Review by stephentoub — APPROVED** [review]
Thanks.  Are there any similar gains to be had in {u}long.TryParse?

**Comment by jeffhandley** [comment]
@GrabYourPitchforks Is this ready to merge?

**Comment by jeffhandley** [comment]
> @GrabYourPitchforks Is this ready to merge?

Ping on this, @GrabYourPitchforks. It would be great to get this in for Preview 7.

**Comment by GrabYourPitchforks** [comment]
I investigated the framework's other parsing routines, but they have a bit more complexity, needing to handle whitespace and trailing nulls. Opened https://github.com/dotnet/runtime/issues/55541 so we don't lose track of this.
